### PR TITLE
CDAP-1965: Provide AbstractCubeHttpHandler out of the box to query a Cube dataset

### PR DESCRIPTION
**Pull request by abaranau**
NOTE: Currently, we can only provide an abstract http handler which subclass has to implement method that returns Cube dataset. Because there's no support of instantiating dataset dynamically based on name in a program.

### CODE REVIEW

**Comment by abaranau** [comment]
http://builds.cask.co/browse/CDAP-DUT1448-1

**Comment by abaranau** [comment]
http://builds.cask.co/browse/CDAP-DUT1448-2

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+  /**
+   * Adds facts to a {@link Cube} as defined by {@link Cube#add(java.util.Collection)}.
+   */
+  @Path("add")
+  @POST
+  public void addFact(HttpServiceRequest request, HttpServiceResponder responder) {
+    try {
+      String body = Charsets.UTF_8.decode(request.getContent()).toString();
+      Collection<CubeFact> facts = new Gson().fromJson(body, new TypeToken<Collection<CubeFact>>() { }.getType());
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: Make the new TypeToken<Collection<CubeFact>>() { }.getType() a constant.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    } catch (Throwable th) {
+      LOG.error("Error while executing request", th);
+      responder.sendError(500, th.getMessage());
+    }
+  }
+
+  /**
+   * Searches tags in a {@link Cube} as defined by {@link Cube#findNextAvailableTags(CubeExploreQuery)}.
+   */
+  @Path("searchTag")
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: Maybe `searchTags`?

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  @POST
+  public void searchTag(HttpServiceRequest request, HttpServiceResponder responder) {
+    try {
+      String body = Charsets.UTF_8.decode(request.getContent()).toString();
+      CubeExploreQuery query = new Gson().fromJson(body, CubeExploreQuery.class);
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: Make the `Gson` a constant.

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+  protected abstract Cube getCube();
+
+  /**
+   * Adds facts to a {@link Cube} as defined by {@link Cube#add(java.util.Collection)}.
+   */
+  @Path("add")
+  @POST
+  public void addFact(HttpServiceRequest request, HttpServiceResponder responder) {
+    try {
+      String body = Charsets.UTF_8.decode(request.getContent()).toString();
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: Maybe just `Bytes.toString(request.getContent());`

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+      responder.sendError(500, th.getMessage());
+    }
+  }
+
+  /**
+   * Searches measurements in a {@link Cube} as defined by {@link Cube#findMeasureNames(CubeExploreQuery)}.
+   */
+  @Path("searchMeasure")
+  @POST
+  public void searchMetric(HttpServiceRequest request, HttpServiceResponder responder) {
```
Comment (gsps1) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: rename method name to ``searchMeasure``

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+import com.google.gson.reflect.TypeToken;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import javax.ws.rs.POST;
+import javax.ws.rs.Path;
+
+/**
+ * A basic implementation of {@link co.cask.cdap.api.service.http.HttpServiceHandler} that provides endpoints to
```
Comment (gsps1) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: this is not implementation of ``HttpServiceHandler`` right ?

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestAppWithCube.java [hunk 7]
```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright © 2014 Cask Data, Inc.
```
Comment (gsps1) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestAppWithCube.java]: 2015

**Comment by gsps1** [comment]
minor comments

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java (hunk 6 above)
Comment (abaranau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: it is, no?

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java (hunk 1 above)
Comment (abaranau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: done

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java (hunk 2 above)
Comment (abaranau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: don't like plural here... You think it should be plural?

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java (hunk 3 above)
Comment (abaranau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: done

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java (hunk 4 above)
Comment (abaranau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: done

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java (hunk 5 above)
Comment (abaranau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: done

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java (hunk 6 above)
Comment (abaranau) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: AbstractHttpServiceHandler implements HttpServiceHandler

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestAppWithCube.java (hunk 7 above)
Comment (abaranau) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestAppWithCube.java]: done

File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java (hunk 2 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/cube/AbstractCubeHttpHandler.java]: No sure, but the response is always an array.

**Comment by abaranau** [comment]
addressed comments

**Comment by gsps1** [comment]
LGTM

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestAppWithCube.java [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+      serviceManager.waitForStatus(true);
+      URL url = serviceManager.getServiceURL();
+
+      long tsInSec = System.currentTimeMillis() / 1000;
+
+      // round to a minute for testing minute resolution
+      tsInSec = (tsInSec / 60) * 60;
+
+      // add couple facts
+      add(url, ImmutableList.of(new CubeFact(ImmutableMap.of("user", "alex", "action", "click"),
```
Comment (anwar6953) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestAppWithCube.java]: Weren't you going to remove your name?

Comment (chtyim) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestAppWithCube.java]: I think "alex" is a very common name.

Comment (abaranau) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/TestAppWithCube.java]: oh, alright. It went it. It's not my name

**Pull request merged**